src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +30,162 @@
 
   private static final String FILTER_ARRAY = "clicked-array";
 
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    HashSet<String> userPlaces = new HashSet();
+    HashSet<String> userDifficulties = new HashSet();
+
+    ArrayList<Destination> allDestinations = new ArrayList();
+    // ArrayList<Name of Destination>
+    ArrayList<String> filteredDestinations = new ArrayList();
+
     // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    allDestinations = createFakeDestinations();
+    userPlaces = separateFilters(clickedFilters, allPlaces);
+    userDifficulties = separateFilters(clickedFilters, allDifficulties);
+    filteredDestinations = filterPlaces(userPlaces, allDestinations);
+    filteredDestinations =
+        filterDifficulty(filteredDestinations, userDifficulties, allDestinations);
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
     response.getWriter().println(clickedFilters);
   }
+
+  public HashSet<String> separateFilters(HashSet<String> clickedFilters, String[] allTypeFilters) {
```

Comment:
Is this essentially taking the two inputs and returning the values that appear in both? I think that behavior can be implemented more simply by using [Set.retainAll()](https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#retainAll(java.util.Collection)). Because that method takes a `Collection` rather than a primitive array, you can convert the input array to a `List` (which is a sub-interface of `Collection`) with [Arrays.asList()](https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html#asList(T...)), either in this method or by calling asList() when initializing the `allPlaces` and `allDifficulties` values at the top.

If you use Set.retainAll(), you may decide to remove this method and do that call directly from the places where this method is currently used, or you may decide it's more readable to keep this separate method. If you do keep this separate method, could you modify the name and/or add a Javadoc to more specifically explain what it does? e.g. `sanitizeFilters` or `restrictFiltersToKnownValues`.

Reply:
+1 to Tim's comment, I think you can remove this method altogether